src/compiler/scala/tools/nsc/transform/Fields.scala

```diff
@@ -412,58 +495,123 @@ abstract class Fields extends InfoTransform with ast.TreeDSL with TypingTransfor
+
+    /**
+      * Desugar a local `lazy val x: Int = rhs` into
+      * ```
+      * val x$lzy = new scala.runtime.LazyInt()
+      * def x(): Int =
+      *   x$lzy.synchronized {
+      *     if (!x$lzy.initialized) {
+      *       x$lzy.initialized = true
+      *       x$lzy.value = rhs
+      *     }
+      *     x$lzy.value
+      *  }
+      * ```
+      */
+    private def mkLazyLocalDef(lazyVal: Symbol, rhs: Tree): Tree = {
+      val lazyValType = lazyVal.tpe.resultType
+      val refClass = lazyHolders.getOrElse(lazyValType.typeSymbol, LazyRefClass)
+      val refTpe = if (refClass != LazyRefClass) refClass.tpe else appliedType(refClass.typeConstructor, List(lazyValType))
+
+      val flags = (lazyVal.flags & FieldFlags | ARTIFACT | MUTABLE) & ~(IMPLICIT | STABLE) // TODO: why include MUTABLE???
+      val name  = lazyVal.name.toTermName.append(nme.LAZY_LOCAL_SUFFIX_STRING)
+      val holderSym =
+        lazyVal.owner.newValue(name, lazyVal.pos, flags) setInfo refTpe
+
+      val accessor = mkAccessor(lazyVal) {
+        import CODE._
+        val initializedGetter = refTpe.member(nme.initialized)
+        val setInitialized    = Apply(Select(Ident(holderSym), initializedGetter.setterIn(refClass)), TRUE :: Nil)
+        val isUnit = refClass == LazyUnitClass
+        val valueGetter = if (isUnit) NoSymbol else refTpe.member(nme.value)
+        val valueSetter = if (isUnit) NoSymbol else valueGetter.setterIn(refClass)
+        val setValue    = if (isUnit) rhs      else Apply(Select(Ident(holderSym), valueSetter), rhs :: Nil)
+        val getValue    = if (isUnit) UNIT     else Apply(Select(Ident(holderSym), valueGetter), Nil)
+
+        // must read the value within the synchronized block since it's not volatile
+        // (there's no happens-before relation with the read of the volatile initialized field)
+        // TODO: double-checked locking (https://github.come/scala/scala-dev/issues/204)
+        gen.mkSynchronized(Ident(holderSym))(
+          If(Ident(holderSym) DOT initializedGetter, getValue, Block(List(setValue, setInitialized), getValue)))
+      }
 
-    def mkField(sym: Symbol) = localTyper.typedPos(sym.pos)(ValDef(sym)).asInstanceOf[ValDef]
+      // do last!
+      // remove LAZY: prevent lazy expansion in mixin
+      // remove STABLE: prevent replacing accessor call of type Unit by BoxedUnit.UNIT in erasure
+      // remove ACCESSOR: prevent constructors from eliminating the method body if the lazy val is
+      // lifted into a trait (TODO: not sure about the details here)
+      lazyVal.resetFlag(LAZY | STABLE | ACCESSOR)
 
+      Thicket(mkField(holderSym, New(refTpe)) :: accessor :: Nil)
+    }
 
     // synth trees for accessors/fields and trait setters when they are mixed into a class
     def fieldsAndAccessors(clazz: Symbol): List[ValOrDefDef] = {
-      def fieldAccess(accessor: Symbol): Option[Tree] = {
+      def fieldAccess(accessor: Symbol): List[Tree] = {
```

Comment:
Why did the result types of `fieldAccess`, `getterBody` and `setterBody` get changed from `Option` to `List`? It seems only empty and one-element lists are constructed.